discojs/src/client/federated/federated_client.ts

```diff
@@ -0,0 +1,247 @@
+   * Disconnection process when user quits the task.
+   */
+  override async disconnect(): Promise<void> {
+    const msg: ClientDisconnected = {
+      type: type.ClientDisconnected,
+    };
+    this.server.send(msg); // notify the server we are disconnecting but don't wait for an answer
+    await this.server.disconnect();
+    this._server = undefined;
+    this._ownId = undefined;
+
+    this.aggregator.setNodes(this.aggregator.nodes.delete(FederatedClient.SERVER_NODE_ID));
+
+    return Promise.resolve();
+  }
+
+  override onRoundBeginCommunication(): Promise<void> {
+    // Prepare the result promise for the incoming round
+    this.aggregationResult = new Promise((resolve) => this.aggregator.once('aggregation', resolve))
+    return Promise.resolve();
+  }
+
+  /**
+   * 
+   * @param weights Local weights sent to the server at the end of the local training round
+   * @param _round The trainer's aggregation round, which can be different from 
+   * the server's round if the participant joined late
+   * @returns the new global weights sent by the server
+   */
+  override async onRoundEndCommunication(weights: WeightsContainer): Promise<WeightsContainer> {
+    // NB: For now, we suppose a fully-federated setting.
+
+    if (this.aggregationResult === undefined) {
+      throw new Error("local aggregation result was not set");
+    }
+
+    // First we check if we are waiting for more participants before sending our weight update
+    if (this.#waitingForMoreParticipants) {
+      if (this.#promiseForMoreParticipants === undefined) throw new Error("Promise should not be undefined")
+        
+      // wait for the promise to resolve, which takes as long as it takes for new participants to join
+      debug("Awaiting the promise for more participants")
+      this.logger.setStatus("Waiting for more participants")
+      await this.#promiseForMoreParticipants 
+      debug("Promise resolved")
+    }
+    this.logger.setStatus("Updating the model with other participants' models")
+
+    // Send our local contribution to the server
+    // and receive the most recent weights as an answer to our contribution
+    const payload: WeightsContainer = this.aggregator.makePayloads(weights).first()
+    const msg: messages.SendPayload = {
+      type: type.SendPayload,
+      payload: await serialization.weights.encode(payload),
+      round: this.aggregator.round,
+    };
+
+    // Need to await the resulting global model right after sending our local contribution
+    // to make sure we don't miss it
+    this.server.send(msg);
+    const serverResult = await this.receiveServerResult();
+
+    if (
+      serverResult !== undefined &&
+      this.aggregator.add(FederatedClient.SERVER_NODE_ID, serverResult, this.aggregator.round)
+    ) {
+      // Regular case: the server sends us its aggregation result which will serve our
+      // own aggregation result.
+    } else {
+      // Unexpected case: for some reason, the server result is stale.
```

Comment:
So far we were just timing out and continuing to the next round which I think makes sense if we expect the server to recover at some point. Maybe we could instead wait indefinitely until the server recovers (and display a status message in the UI) ? Failing would also make sense after waiting long enough

Reply:
yeah, waiting indefinitely (or a looong time) makes more sense to me. if the server crashes, I'm not even sure that it can restart a training. and the newer round will get rejected by the server anyway.